common/lib/client/realtimechannel.js

```diff
@@ -409,68 +423,47 @@ var RealtimeChannel = (function() {
-				this.setState('detached');
-			}
-			this.failPendingMessages(new ErrorInfo('Channel detached', 90001, 404));
+		if(state === this.state) {
+			return;
 		}
-	};
-
-	RealtimeChannel.prototype.setFailed = function(message) {
-		this.clearStateTimer();
-		var err = ErrorInfo.fromValues(message.error || {statusCode: 400, code: 90000, message: 'Channel failed'});
-		this.setState('failed', err);
-		this.failPendingMessages(err);
-	};
-
-	RealtimeChannel.prototype.setSuspended = function(err, suppressEvent) {
-		if(this.state !== 'detached' && this.state !== 'failed') {
-			Logger.logAction(Logger.LOG_MINOR, 'RealtimeChannel.setSuspended', 'deactivating channel; name = ' + this.name + ', err ' + (err ? err.message : 'none'));
-			this.clearStateTimer();
-			this.presence.setSuspended(err);
-			if(!suppressEvent) {
-				this.setState('detached');
-			}
-			this.failPendingMessages(err);
+    if(state !== 'attached' && state !== 'attaching') {
+      this.presence.setSuspended(reason || RealtimeChannel.invalidStateError);
+      this.failPendingMessages(reason || RealtimeChannel.invalidStateError);
+    }
+		if(state === 'suspended' && this.connectionManager.state.sendEvents) {
+			this.startRetryTimer();
+		} else {
+			this.cancelRetryTimer();
 		}
-	};
-
-	RealtimeChannel.prototype.setState = function(state, err) {
-		this.state = state;
-		if(err) {
-			this.errorReason = err;
+		if(reason) {
+			this.errorReason = reason;
 		}
+    var change = new ChannelStateChange(this.state, state, false /* TODO resumed */, reason);
 		var logLevel = state === 'failed' ? Logger.LOG_ERROR : Logger.LOG_MAJOR;
-		Logger.logAction(logLevel, 'Channel state for channel "' + this.name + '"', state + (err ? ('; reason: ' + err.message + ', code: ' + err.code) : ''));
+		Logger.logAction(logLevel, 'Channel state for channel "' + this.name + '"', state + (reason ? ('; reason: ' + reason.toString()) : ''));
+
+		/* Note: we don't set inProgress for pending states until the request is actually in progress */
 		if(state === 'attached') {
 			this.setInProgress(statechangeOp, false);
-		} else if(state === 'detached' || state === 'failed') {
+		} else if(state === 'detached' || state === 'failed' || state === 'suspended') {
 			this.setInProgress(statechangeOp, false);
 			this.setInProgress(syncOp, false);
 		}
-		this.emit(state, err);
-	};
 
-	RealtimeChannel.prototype.setPendingState = function(state) {
-		Logger.logAction(Logger.LOG_MINOR, 'RealtimeChannel.setPendingState', 'name = ' + this.name + ', state = ' + state);
-		this.clearStateTimer();
-
-		/* notify the state change, but don't set inProgress until the request is actually in progress */
-		this.setState(state);
+		this.state = state;
+		this.emit(state, change);
+	};
 
+	RealtimeChannel.prototype.requestState = function(state, reason) {
+		Logger.logAction(Logger.LOG_MINOR, 'RealtimeChannel.requestState', 'name = ' + this.name + ', state = ' + state);
+    this.notifyState(state, reason);
```

Comment:
all this stuff was already fixed in later commits. IIRC I tried to squash it down, but it didn't squash cleanly, didn't bother spending too much time working out why